test/integration/controller/jobset_controller_test.go

```diff
@@ -2498,6 +2499,73 @@ var _ = ginkgo.Describe("JobSet controller", func() {
 			}, timeout, interval).Should(gomega.BeTrue())
 		})
 	})
+
+	ginkgo.When("a JobSet is created with an invalid Job template", func() {
+		ginkgo.It("should emit a warning event", func() {
+			ns := &corev1.Namespace{
+				ObjectMeta: metav1.ObjectMeta{
+					GenerateName: "jobset-ns-",
+				},
+			}
+
+			ginkgo.By("creating namespace")
+			gomega.Expect(k8sClient.Create(ctx, ns)).To(gomega.Succeed())
+
+			defer func() {
+				gomega.Expect(testutil.DeleteNamespace(ctx, k8sClient, ns)).To(gomega.Succeed())
+			}()
+
+			ginkgo.By("creating jobset with an invalid job template")
+			podSpec := testing.TestPodSpec
+			podSpec.Containers[0].Image = ""
+
+			js := testing.MakeJobSet("invalid-jobset", ns.Name).
+				ReplicatedJob(testing.MakeReplicatedJob("replicated-job-a").
+					Job(testing.MakeJobTemplate("test-job-A", ns.Name).PodSpec(podSpec).Obj()).
+					Replicas(1).
+					Obj()).
+				Obj()
+
+			gomega.Expect(k8sClient.Create(ctx, js)).Should(gomega.Succeed())
+
+			ginkgo.By("checking that a warning event is emitted")
+			gomega.Eventually(func() (bool, error) {
+				var events corev1.EventList
+				if err := k8sClient.List(ctx, &events, client.InNamespace(ns.Name), client.MatchingFieldsSelector{
+					Selector: fields.AndSelectors(
+						fields.OneTermEqualSelector("involvedObject.kind", "JobSet"),
+						fields.OneTermEqualSelector("involvedObject.name", js.Name),
+					),
+				}); err != nil {
+					return false, err
+				}
+
+				if len(events.Items) < 1 {
+					return false, nil
+				}
+
+				event := events.Items[0]
+				gomega.Expect(event.Type).To(gomega.Equal(corev1.EventTypeWarning))
+				gomega.Expect(event.Reason).To(gomega.Equal("JobCreationFailed"))
+				return true, nil
+			}, timeout, interval).Should(gomega.BeTrue())
+
+			ginkgo.By("checking that the events are accumulated")
```

Comment:
Strictly speaking the following test doesn't check there is just a single API request to add the events, but the correlation broadcaster setup can be seen here: https://github.com/kubernetes/client-go/blob/v0.34.1/tools/record/event.go#L276

Reply:
Ensuring this properly would require mocking the `EventSink` and creating an `EvenBroadcaster` manually:
https://pkg.go.dev/k8s.io/client-go/tools/record#EventSink